[ZB-6720] Modify Tform to trace output keys to input keys

jjmartin14 opened the pull request:
## Description
- Add `_tracePropAccess`, which wraps a given object with a Proxy function that traces which of its properties are accessed by other functions
- In `_processRules`, pass `record` wrapped in `_tracePropAccess` in addition to just `record`
- Add `getKeyMapAfterTransform`, which, after a call to `transform`, will contain a mapping of output keys to the input keys their corresponding rules accessed

**NOTE: `getKeyMapAfterTransform` will only contain the input keys actually accessed by the rules, so if there are conditional statements, it may not contain all of the possible input keys.**

**NOTE 2: if there are bottom-level keys with the same output key name, the mapping for that name will go to a set with all of their input keys. This mirrors the behavior of `_addErrors`, which lists all errors for one output key name under one label.**

**NOTE 3: `getKeyMapAfterTransform` will return an empty map if `transform` is not called first.**

## Test Plan
- Added tests to `tform-test`
- The full map object for the `basic transforming` test in `tform-test` is:
<img width="720" alt="Screen Shot 2019-08-12 at 4 37 27 PM" src="https://user-images.githubusercontent.com/51676038/62905481-913f6e80-bd1f-11e9-9298-aebd568e3a59.png">
(can update the test to check everything if we would prefer. would just be adding a lot of lines)

## Relevant Tasks
[ZB-6720]

## Post Merge Plan

jjmartin14 committed:
Add auditing functionality and tests for it

jjmartin14 committed:
Do not overwrite existing mappings while traversing recursive rules

jlw5 reviewed (commented):
Note 1 sounds like the right behavior. If a given record only uses 1 of 2 possible (conditional) properties and fails, then we only want to blame the 1 that it used.

This looks like it will work. Did you decide on modifying Tform vs bringing it into ImportHandler?

jjmartin14 commented:
I think modifying Tform is better if it's ok! It seems inefficient to basically duplicate Tform in ImportHandler if people are ok with changing Tform.

jf248 commented:
@jjmartin14 have you measured the performance implication of wrapping each object in a Proxy?

jf248 commented:
@jjmartin14 ^ please remember, as I'm sure you will, that this repo is public so make sure to only comment here generically not referring to any specific Rimeto implementations

jjmartin14 commented:
@jf248 

I modified `index` (without pushing to this PR) to have a timer (in milliseconds):

![Screen Shot 2019-08-12 at 5 32 58 PM](https://user-images.githubusercontent.com/51676038/62907278-49244a00-bd27-11e9-871f-eaa0ec1d0039.png)

Output from `yarn test`:
<img width="411" alt="Screen Shot 2019-08-12 at 5 55 10 PM" src="https://user-images.githubusercontent.com/51676038/62907923-5abb2100-bd2a-11e9-83ab-734b44f863a9.png">

Output from `yarn test -t "basic transforming"`:
<img width="566" alt="Screen Shot 2019-08-12 at 5 33 54 PM" src="https://user-images.githubusercontent.com/51676038/62907932-63135c00-bd2a-11e9-8558-66e85e45f241.png">

Output from `yarn test -t "basic error handling"` (smaller input record than "basic transforming"):
<img width="698" alt="Screen Shot 2019-08-12 at 5 55 53 PM" src="https://user-images.githubusercontent.com/51676038/62907947-76262c00-bd2a-11e9-90cf-fd609c1f7e47.png">

It looks like the first instance of the wrapping is the most expensive, at almost 1/10th of a millisecond, and additional calls are on the order of 1/100th of a millisecond. So, using this method (wrapping the object in Proxy for each rule) could get pretty time-consuming with more rules, but a large chunk of time will be from doing it at all.

(I'm not sure at what points in use the first, most expensive call would have to be redone. It doesn't happen for new input records or rules, as evidenced by the tests which use multiple records/rules.)

jjmartin14 committed:
Wrap once for all rules instead of for every rule; rename function for getting map

jjmartin14 commented:
Updated so that the wrapping is done only once per `transform`, instead of once per `rule`.

Also renamed the function for getting the key map to `getKeyMapAfterTransform` to try to emphasize it should be called after `transform`.

jjmartin14 commented:
Changed to our own local version of Tform instead since we want to test more specifically and move away from supporting Tform.
